### PR TITLE
LWDEV-10470 Add notification throttling functionality to trade volume service

### PR DESCRIPTION
**Pull request by olpapchenko**

### CODE REVIEW

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class SentNotificationsCacheImpl(config: Config): SentNotificationsCache {
+    private val notificationConfig = config.tradeVolumeConfig.notificationsConfig
+    private val timestampByNotificationKey = ConcurrentHashMap<String, Date>()
+
+    override fun add(clientId: String, assetId: String) {
+        timestampByNotificationKey[getKey(clientId, assetId)] = Date()
+    }
+
+    override fun isSent(clientId: String, assetId: String): Boolean {
+        return timestampByNotificationKey[getKey(clientId, assetId)] != null
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt]: Compare also `timestampByNotificationKey[getKey(clientId, assetId)]` (if it is not `null`) with `(Date().time - notificationConfig.throttlingPeriod`

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    @Scheduled(fixedRateString = "#{Config.tradeVolumeConfig.notificationsConfig.throttlingPeriod}")
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt:24]: `fixedRateString` should be get from another setting. `throttlingPeriod` is lifetime of record about sent notification. But we need frequency of cleaning here

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package com.lykke.trade.volume.monitoring.service.notification
+
+import com.lykke.trade.volume.monitoring.service.config.Config
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+
+@Component
+class SentNotificationsCacheImpl(config: Config): SentNotificationsCache {
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt]: It is better to pass only used parameter values to bean but not whole config

**Review by barsukovskii — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt (hunk 1 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt]: fixed

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt (hunk 2 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt:24]: agreed to leave one parameter

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt (hunk 3 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/SentNotificationsCacheImpl.kt]: fixed

**Review by barsukovskii — APPROVED** [review]

**Pull request merged**